agent/exec/containerd/adapter.go

```diff
@@ -0,0 +1,682 @@
+	imageStore := imagesservice.NewStoreFromClient(imagesapi.NewImagesClient(c.conn))
+
+	image, err := imageStore.Get(ctx, c.resolvedImageName)
+	if err != nil {
+		return errors.Wrap(err, "Image Get")
+	}
+
+	mbytes, err := content.ReadBlob(ctx, cs, image.Target.Digest)
+	if err != nil {
+		return err
+	}
+
+	rootfs := filepath.Join(c.dir, "rootfs")
+	// TODO(ijc) support ControllerLogs interface
+	stdin := "/dev/null"
+	stdout := filepath.Join(c.dir, "stdout")
+	stderr := filepath.Join(c.dir, "stderr")
+
+	if err := os.MkdirAll(rootfs, 0755); err != nil {
+		return err
+	}
+
+	var config ocispec.Image
+
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(mbytes, &manifest); err != nil {
+		return err
+	}
+
+	bytes, err := content.ReadBlob(ctx, cs, manifest.Config.Digest)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(bytes, &config); err != nil {
+		return err
+	}
+
+	for _, layer := range manifest.Layers {
+		if err := c.applyLayer(ctx, cs, rootfs, layer.Digest); err != nil {
+			return errors.Wrapf(err, "Failed to apply layer %s", layer.Digest.String())
+		}
+	}
+
+	spec, err := c.spec(ctx, &config.Config, rootfs)
+	if err != nil {
+		return err
+	}
+
+	_, err = prepareStdio(stdout, stderr, spec.Process.Terminal)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.MarshalIndent(spec, "    ", "    ")
+	if err != nil {
+		return err
+	}
+
+	_, err = c.client.Create(ctx, &execution.CreateRequest{
+		ID: naming.Task(c.task),
+		Spec: &protobuf.Any{
+			TypeUrl: specs.Version,
+			Value:   data,
+		},
+		Rootfs:   []*mount.Mount{},
+		Runtime:  "linux",
+		Stdin:    stdin,
+		Stdout:   stdout,
+		Stderr:   stderr,
+		Terminal: spec.Process.Terminal,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *containerAdapter) start(ctx context.Context) error {
+	_, err := c.client.Start(ctx, &execution.StartRequest{
+		ID: naming.Task(c.task),
+	})
+	return err
+}
+
+func (c *containerAdapter) eventStream(ctx context.Context, id string) (<-chan container.Event, <-chan error, error) {
+
+	var (
+		evtch = make(chan container.Event)
+		errch = make(chan error)
+	)
+
+	return evtch, errch, nil
+}
+
+// events issues a call to the events API and returns a channel with all
+// events. The stream of events can be shutdown by cancelling the context.
+//
+// A chan struct{} is returned that will be closed if the event processing
+// fails and needs to be restarted.
+func (c *containerAdapter) events(ctx context.Context) (<-chan container.Event, <-chan struct{}, error) {
+	id := naming.Task(c.task)
+
+	l := log.G(ctx).WithFields(logrus.Fields{
+		"ID": id,
+	})
+
+	// TODO(stevvooe): Move this to a single, global event dispatch. For
+	// now, we create a connection per container.
+	var (
+		eventsq = make(chan container.Event)
+		closed  = make(chan struct{})
+	)
+
+	l.Debugf("waiting on events")
+	// TODO(stevvooe): For long running tasks, it is likely that we will have
+	// to restart this under failure.
```

Comment:
Then again, if this can only be a Unix socket (as you say below), this might not matter much. The main thing to consider would be the behavior when containerd dies - should the executor treat that as all tasks failing, or just as a temporary loss of the event feed?